File: LeetCode/Top-Interview-150/Divide-and-Conquer/SortedArraytoBST.java
public class TreeNode 
{
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) 
    {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution 
{
    public TreeNode sortedArrayToBST(int[] nums) 
    {
        if (nums == null || nums.length == 0) 
        {
            return null;
        }

        return constructBSTRecursive(nums, 0, nums.length - 1);
    }
    
    private TreeNode constructBSTRecursive(int[] nums, int start, int end) 
    {
        if (start > end) 
        {
            return null;
        }
        
        int mid = start + (end - start) / 2;
        TreeNode root = new TreeNode(nums[mid]);
        
        root.left = constructBSTRecursive(nums, start, mid - 1);
        root.right = constructBSTRecursive(nums, mid + 1, end);
        
        return root;
    }
}

/*

Time Complexity:

- O(n)
- Each element is traversed once

 */

File: LeetCode/Top-Interview-150/Divide-and-Conquer/SortedArrayToBST.py
# Definition for a binary tree node.
# class TreeNode:
#    def __init__(self, val=0, left=None, right=None):
#        self.val = val
#        self.left = left
#        self.right = right

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        
        if not nums:
            return None

        # The middle element will be the root in the sorted array
        mid = len(nums) // 2

        # Create a new tree node with the root
        root = TreeNode(nums[mid])

        # Left subtree
        root.left = self.sortedArrayToBST(nums[:mid])

        # Right subtree
        root.right = self.sortedArrayToBST(nums[mid+1:])

        return root
    
'''

Time Complexity:

- O(n)
- Each element is traversed once

'''